sql/sql_tablespace.cc

```diff
@@ -932,6 +934,44 @@ bool Sql_cmd_alter_tablespace::execute(THD *thd) {
     return true;
   }
 
+  class scope_defer {
+    void (*func)(void) = nullptr;
+
+   public:
+    ~scope_defer() {
+      if (func) func();
+    }
+    void reset(void (*new_func)(void)) { func = new_func; }
+  } scope_defer;
+
+  if (m_tablespace_name.length == MYSQL_TABLESPACE_NAME.length &&
+      memcmp(m_tablespace_name.str, MYSQL_TABLESPACE_NAME.str,
+             MYSQL_TABLESPACE_NAME.length) == 0) {
+    /*
+      "mysql" tablespace is a special case, as data dictionary tables (residing
+      in "mysql") are expected to be always readable without creating deadlocks.
+      Exclusive locking of tables in "mysql" tablespace further down this
+      function opens a small window of opportunity to cause potential deadlocks
+      in dictionary tables updates from concurrent threads. Which in turn may
+      cause outer transactions to roll back, which will break no-rollback
+      guarantees of (nested) attachable transactions of data dictionary
+      accesses.
+
+      To prevent that, only continue with encryption of "mysql" tablespace if
+      this thread is the only client connected. Encryption of other tablespaces
+      is not affected.
+    */
+    scope_defer.reset(
+        []() { Connection_handler_manager::resume_new_connections(); });
+
+    Connection_handler_manager::pause_new_connections();
+    if (Connection_handler_manager::connection_count > 1) {
+      my_error(ER_CHECK_NOT_IMPLEMENTED, MYF(0),
+               "changing ENCRYPTION state while other clients connected");
+      return true;
+    }
+  }
+
```

Comment:
Good observation. The attachable trxs should never be rolled back. Can we fix this issue as separate PR?
Also, can you please provide me stack trace of attachable trxs failing (or asserting?)

Reply:
Another option is to fix mysql tablesapce encryption-decryption before we accept connections.

Reply:
> stack trace of attachable trxs failing

Here they are. Click to expand.
<details><summary>Stack track at the point of assertion:</summary>

```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x0000038f69eea859 in __GI_abort () at abort.c:79
#2  0x0000038f69eea729 in __assert_fail_base (fmt=0x38f6a080588 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=0x56304b9b0050 "!m_thd->transaction_rollback_request", file=0x56304b9aff90 "/work/percona-server/sql/sql_class.cc", line=270, function=<optimized out>) at assert.c:92
#3  0x0000038f69efbf36 in __GI___assert_fail (assertion=0x56304b9b0050 "!m_thd->transaction_rollback_request", file=0x56304b9aff90 "/work/percona-server/sql/sql_class.cc", line=270, function=0x56304b9b0020 "virtual THD::Attachable_trx::~Attachable_trx()") at assert.c:101
#4  0x000056304dd7adb7 in THD::Attachable_trx::~Attachable_trx (this=0x78b438295840, __in_chrg=<optimized out>) at /work/percona-server/sql/sql_class.cc:270
#5  0x000056304dd7aeb0 in THD::Attachable_trx::~Attachable_trx (this=0x78b438295840, __in_chrg=<optimized out>) at /work/percona-server/sql/sql_class.cc:306
#6  0x000056304dd81c76 in THD::end_attachable_transaction (this=0x78b4380cd3f0) at /work/percona-server/sql/sql_class.cc:2060
#7  0x000056304f0a5ff0 in dd::Transaction_ro::~Transaction_ro (this=0x2ae05403ff30, __in_chrg=<optimized out>) at /work/percona-server/sql/dd/impl/transaction_impl.h:138
#8  0x000056304f1bfe8f in dd::cache::Storage_adapter::get<dd::Item_name_key, dd::Abstract_table> (thd=0x78b4380cd3f0, key=..., isolation=ISO_READ_COMMITTED, bypass_core_registry=false, object=0x2ae054040040) at /work/percona-server/sql/dd/impl/cache/storage_adapter.cc:167
#9  0x000056304f1ba278 in dd::cache::Shared_dictionary_cache::get_uncached<dd::Item_name_key, dd::Abstract_table> (this=0x5630506e7ea0 <dd::cache::Shared_dictionary_cache::instance()::s_cache>, thd=0x78b4380cd3f0, key=..., isolation=ISO_READ_COMMITTED, object=0x2ae054040040) at /work/percona-server/sql/dd/impl/cache/shared_dictionary_cache.cc:113
#10 0x000056304f1ba021 in dd::cache::Shared_dictionary_cache::get<dd::Item_name_key, dd::Abstract_table> (this=0x5630506e7ea0 <dd::cache::Shared_dictionary_cache::instance()::s_cache>, thd=0x78b4380cd3f0, key=..., element=0x2ae0540400a8) at /work/percona-server/sql/dd/impl/cache/shared_dictionary_cache.cc:98
#11 0x000056304f0c81ba in dd::cache::Dictionary_client::acquire<dd::Item_name_key, dd::Abstract_table> (this=0x78b438006700, key=..., object=0x2ae054040138, local_committed=0x2ae05404012d, local_uncommitted=0x2ae05404012e) at /work/percona-server/sql/dd/impl/cache/dictionary_client.cc:909
#12 0x000056304f0a8786 in dd::cache::Dictionary_client::acquire<dd::Abstract_table> (this=0x78b438006700, schema_name="mtr", object_name="global_suppressions", object=0x2ae054040278) at /work/percona-server/sql/dd/impl/cache/dictionary_client.cc:1378
#13 0x000056304dd4efeb in get_table_share (thd=0x78b4380cd3f0, db=0x78b4381e2620 "mtr", table_name=0x78b4381ec118 "global_suppressions", key=0x78b4381ee617 "mtr", key_length=24, open_view=true, open_secondary=false) at /work/percona-server/sql/sql_base.cc:767
#14 0x000056304dd4f7b2 in get_table_share_with_discover (thd=0x78b4380cd3f0, table_list=0x78b4381ee240, key=0x78b4381ee617 "mtr", key_length=24, open_secondary=false, error=0x2ae0540404a4) at /work/percona-server/sql/sql_base.cc:883
#15 0x000056304dd55595 in open_table (thd=0x78b4380cd3f0, table_list=0x78b4381ee240, ot_ctx=0x2ae054040980) at /work/percona-server/sql/sql_base.cc:3222
#16 0x000056304dd59c16 in open_and_process_table (thd=0x78b4380cd3f0, lex=0x78b4381ea770, tables=0x78b4381ee240, counter=0x78b4381ea7c8, prelocking_strategy=0x2ae054040a08, has_prelocking_list=false, ot_ctx=0x2ae054040980) at /work/percona-server/sql/sql_base.cc:5085
#17 0x000056304dd5b860 in open_tables (thd=0x78b4380cd3f0, start=0x2ae0540409f0, counter=0x78b4381ea7c8, flags=0, prelocking_strategy=0x2ae054040a08) at /work/percona-server/sql/sql_base.cc:5895
#18 0x000056304dd5d3d4 in open_tables_for_query (thd=0x78b4380cd3f0, tables=0x78b4381ed7f8, flags=0) at /work/percona-server/sql/sql_base.cc:6775
#19 0x000056304df80370 in Sql_cmd_dml::prepare (this=0x78b4381ef360, thd=0x78b4380cd3f0) at /work/percona-server/sql/sql_select.cc:361
#20 0x000056304df80db9 in Sql_cmd_dml::execute (this=0x78b4381ef360, thd=0x78b4380cd3f0) at /work/percona-server/sql/sql_select.cc:521
#21 0x000056304def1611 in mysql_execute_command (thd=0x78b4380cd3f0, first_level=false) at /work/percona-server/sql/sql_parse.cc:3641
#22 0x000056304dd00c59 in sp_instr_stmt::exec_core (this=0x78b4381ef458, thd=0x78b4380cd3f0, nextp=0x2ae054042014) at /work/percona-server/sql/sp_instr.cc:1000
#23 0x000056304dcff611 in sp_lex_instr::reset_lex_and_exec_core (this=0x78b4381ef458, thd=0x78b4380cd3f0, nextp=0x2ae054042014, open_tables=false) at /work/percona-server/sql/sp_instr.cc:443
#24 0x000056304dd0023f in sp_lex_instr::validate_lex_and_execute_core (this=0x78b4381ef458, thd=0x78b4380cd3f0, nextp=0x2ae054042014, open_tables=false) at /work/percona-server/sql/sp_instr.cc:729
#25 0x000056304dd0086b in sp_instr_stmt::execute (this=0x78b4381ef458, thd=0x78b4380cd3f0, nextp=0x2ae054042014) at /work/percona-server/sql/sp_instr.cc:906
#26 0x000056304dcf1f72 in sp_head::execute (this=0x78b4381e1be8, thd=0x78b4380cd3f0, merge_da_on_success=true) at /work/percona-server/sql/sp_head.cc:2223
#27 0x000056304dcf4100 in sp_head::execute_procedure (this=0x78b4381e1be8, thd=0x78b4380cd3f0, args=0x0) at /work/percona-server/sql/sp_head.cc:2875
#28 0x000056304dd79ad1 in Sql_cmd_call::execute_inner (this=0x78b438143b80, thd=0x78b4380cd3f0) at /work/percona-server/sql/sql_call.cc:235
#29 0x000056304df81152 in Sql_cmd_dml::execute (this=0x78b438143b80, thd=0x78b4380cd3f0) at /work/percona-server/sql/sql_select.cc:574
#30 0x000056304def4e76 in mysql_execute_command (thd=0x78b4380cd3f0, first_level=false) at /work/percona-server/sql/sql_parse.cc:4720
#31 0x000056304dd00c59 in sp_instr_stmt::exec_core (this=0x78b438143bf8, thd=0x78b4380cd3f0, nextp=0x2ae054043d84) at /work/percona-server/sql/sp_instr.cc:1000
#32 0x000056304dcff611 in sp_lex_instr::reset_lex_and_exec_core (this=0x78b438143bf8, thd=0x78b4380cd3f0, nextp=0x2ae054043d84, open_tables=false) at /work/percona-server/sql/sp_instr.cc:443
#33 0x000056304dd0023f in sp_lex_instr::validate_lex_and_execute_core (this=0x78b438143bf8, thd=0x78b4380cd3f0, nextp=0x2ae054043d84, open_tables=false) at /work/percona-server/sql/sp_instr.cc:729
#34 0x000056304dd0086b in sp_instr_stmt::execute (this=0x78b438143bf8, thd=0x78b4380cd3f0, nextp=0x2ae054043d84) at /work/percona-server/sql/sp_instr.cc:906
#35 0x000056304dcf1f72 in sp_head::execute (this=0x78b438139cf8, thd=0x78b4380cd3f0, merge_da_on_success=true) at /work/percona-server/sql/sp_head.cc:2223
#36 0x000056304dcf4100 in sp_head::execute_procedure (this=0x78b438139cf8, thd=0x78b4380cd3f0, args=0x78b4380d6960) at /work/percona-server/sql/sp_head.cc:2875
#37 0x000056304dd79ad1 in Sql_cmd_call::execute_inner (this=0x78b4380d6e28, thd=0x78b4380cd3f0) at /work/percona-server/sql/sql_call.cc:235
#38 0x000056304df81152 in Sql_cmd_dml::execute (this=0x78b4380d6e28, thd=0x78b4380cd3f0) at /work/percona-server/sql/sql_select.cc:574
#39 0x000056304def4e76 in mysql_execute_command (thd=0x78b4380cd3f0, first_level=true) at /work/percona-server/sql/sql_parse.cc:4720
#40 0x000056304def7014 in dispatch_sql_command (thd=0x78b4380cd3f0, parser_state=0x2ae054045aa0, update_userstat=false) at /work/percona-server/sql/sql_parse.cc:5325
#41 0x000056304deec4f6 in dispatch_command (thd=0x78b4380cd3f0, com_data=0x2ae054046b70, command=COM_QUERY) at /work/percona-server/sql/sql_parse.cc:1956
#42 0x000056304deea868 in do_command (thd=0x78b4380cd3f0) at /work/percona-server/sql/sql_parse.cc:1404
#43 0x000056304e1145d6 in handle_connection (arg=0x23fb4af16320) at /work/percona-server/sql/conn_handler/connection_handler_per_thread.cc:307
#44 0x000056304fba1862 in pfs_spawn_thread (arg=0x23fb4aea0760) at /work/percona-server/storage/perfschema/pfs.cc:2899
#45 0x00003c74345ce609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#46 0x0000038f69fe7293 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```
</details>

<details><summary>Stack trace at the point where `m_thd->transaction_rollback_request` is set:</summary>

```
#0  0x000056304dd83380 in THD::mark_transaction_to_rollback (this=0x78b4380cd3f0, all=true) at /work/percona-server/sql/sql_class.cc:2504
#1  0x000056304dd57cbf in Open_table_context::request_backoff_action (this=0x2ae05403fdb0, action_arg=Open_table_context::OT_BACKOFF_AND_RETRY, table=0x0) at /work/percona-server/sql/sql_base.cc:4170
#2  0x000056304dd6c6fd in MDL_deadlock_handler::handle_condition (this=0x2ae05403f690, sql_errno=1213) at /work/percona-server/sql/sql_base.cc:2540
#3  0x000056304dd7cc84 in THD::handle_condition (this=0x78b4380cd3f0, sql_errno=1213, sqlstate=0x56304c848673 "40001", level=0x2ae05403f100, msg=0x2ae05403f2b0 "Deadlock found when trying to get lock; try restarting transaction") at /work/percona-server/sql/sql_class.cc:675
#4  0x000056304dd7d606 in THD::raise_condition (this=0x78b4380cd3f0, sql_errno=1213, sqlstate=0x56304c848673 "40001", level=Sql_condition::SL_ERROR, msg=0x2ae05403f2b0 "Deadlock found when trying to get lock; try restarting transaction", fatal_error=false) at /work/percona-server/sql/sql_class.cc:778
#5  0x000056304dbaa164 in my_message_sql (error=1213, str=0x2ae05403f2b0 "Deadlock found when trying to get lock; try restarting transaction", MyFlags=0) at /work/percona-server/sql/mysqld.cc:3895
#6  0x000056304f380726 in my_error (nr=1213, MyFlags=0) at /work/percona-server/mysys/my_error.cc:249
#7  0x000056304db98044 in MDL_context::acquire_lock (this=0x78b4380cd468, mdl_request=0x78b438295228, lock_wait_timeout=31536000) at /work/percona-server/sql/mdl.cc:3567
#8  0x000056304dd5424c in open_table_get_mdl_lock (thd=0x78b4380cd3f0, ot_ctx=0x2ae05403fdb0, table_list=0x78b438294e78, flags=18434, mdl_ticket=0x2ae05403f900) at /work/percona-server/sql/sql_base.cc:2711
#9  0x000056304dd55057 in open_table (thd=0x78b4380cd3f0, table_list=0x78b438294e78, ot_ctx=0x2ae05403fdb0) at /work/percona-server/sql/sql_base.cc:3090
#10 0x000056304dd59c16 in open_and_process_table (thd=0x78b4380cd3f0, lex=0x78b4381ea770, tables=0x78b438294e78, counter=0x2ae05403fe90, prelocking_strategy=0x2ae05403fe30, has_prelocking_list=false, ot_ctx=0x2ae05403fdb0) at /work/percona-server/sql/sql_base.cc:5085
#11 0x000056304dd5b860 in open_tables (thd=0x78b4380cd3f0, start=0x2ae05403fe88, counter=0x2ae05403fe90, flags=18434, prelocking_strategy=0x2ae05403fe30) at /work/percona-server/sql/sql_base.cc:5895
#12 0x000056304dd6a0ee in open_tables (thd=0x78b4380cd3f0, tables=0x2ae05403fe88, counter=0x2ae05403fe90, flags=18434) at /work/percona-server/sql/sql_base.h:462
#13 0x000056304f07a0ba in dd::Open_dictionary_tables_ctx::open_tables (this=0x2ae05403ff30) at /work/percona-server/sql/dd/impl/transaction_impl.cc:107
#14 0x000056304f1bfb4a in dd::cache::Storage_adapter::get<dd::Item_name_key, dd::Abstract_table> (thd=0x78b4380cd3f0, key=..., isolation=ISO_READ_COMMITTED, bypass_core_registry=false, object=0x2ae054040040) at /work/percona-server/sql/dd/impl/cache/storage_adapter.cc:170
#15 0x000056304f1ba278 in dd::cache::Shared_dictionary_cache::get_uncached<dd::Item_name_key, dd::Abstract_table> (this=0x5630506e7ea0 <dd::cache::Shared_dictionary_cache::instance()::s_cache>, thd=0x78b4380cd3f0, key=..., isolation=ISO_READ_COMMITTED, object=0x2ae054040040) at /work/percona-server/sql/dd/impl/cache/shared_dictionary_cache.cc:113
#16 0x000056304f1ba021 in dd::cache::Shared_dictionary_cache::get<dd::Item_name_key, dd::Abstract_table> (this=0x5630506e7ea0 <dd::cache::Shared_dictionary_cache::instance()::s_cache>, thd=0x78b4380cd3f0, key=..., element=0x2ae0540400a8) at /work/percona-server/sql/dd/impl/cache/shared_dictionary_cache.cc:98
#17 0x000056304f0c81ba in dd::cache::Dictionary_client::acquire<dd::Item_name_key, dd::Abstract_table> (this=0x78b438006700, key=..., object=0x2ae054040138, local_committed=0x2ae05404012d, local_uncommitted=0x2ae05404012e) at /work/percona-server/sql/dd/impl/cache/dictionary_client.cc:909
#18 0x000056304f0a8786 in dd::cache::Dictionary_client::acquire<dd::Abstract_table> (this=0x78b438006700, schema_name="mtr", object_name="global_suppressions", object=0x2ae054040278) at /work/percona-server/sql/dd/impl/cache/dictionary_client.cc:1378
#19 0x000056304dd4efeb in get_table_share (thd=0x78b4380cd3f0, db=0x78b4381e2620 "mtr", table_name=0x78b4381ec118 "global_suppressions", key=0x78b4381ee617 "mtr", key_length=24, open_view=true, open_secondary=false) at /work/percona-server/sql/sql_base.cc:767
#20 0x000056304dd4f7b2 in get_table_share_with_discover (thd=0x78b4380cd3f0, table_list=0x78b4381ee240, key=0x78b4381ee617 "mtr", key_length=24, open_secondary=false, error=0x2ae0540404a4) at /work/percona-server/sql/sql_base.cc:883
#21 0x000056304dd55595 in open_table (thd=0x78b4380cd3f0, table_list=0x78b4381ee240, ot_ctx=0x2ae054040980) at /work/percona-server/sql/sql_base.cc:3222
#22 0x000056304dd59c16 in open_and_process_table (thd=0x78b4380cd3f0, lex=0x78b4381ea770, tables=0x78b4381ee240, counter=0x78b4381ea7c8, prelocking_strategy=0x2ae054040a08, has_prelocking_list=false, ot_ctx=0x2ae054040980) at /work/percona-server/sql/sql_base.cc:5085
#23 0x000056304dd5b860 in open_tables (thd=0x78b4380cd3f0, start=0x2ae0540409f0, counter=0x78b4381ea7c8, flags=0, prelocking_strategy=0x2ae054040a08) at /work/percona-server/sql/sql_base.cc:5895
#24 0x000056304dd5d3d4 in open_tables_for_query (thd=0x78b4380cd3f0, tables=0x78b4381ed7f8, flags=0) at /work/percona-server/sql/sql_base.cc:6775
#25 0x000056304df80370 in Sql_cmd_dml::prepare (this=0x78b4381ef360, thd=0x78b4380cd3f0) at /work/percona-server/sql/sql_select.cc:361
#26 0x000056304df80db9 in Sql_cmd_dml::execute (this=0x78b4381ef360, thd=0x78b4380cd3f0) at /work/percona-server/sql/sql_select.cc:521
#27 0x000056304def1611 in mysql_execute_command (thd=0x78b4380cd3f0, first_level=false) at /work/percona-server/sql/sql_parse.cc:3641
#28 0x000056304dd00c59 in sp_instr_stmt::exec_core (this=0x78b4381ef458, thd=0x78b4380cd3f0, nextp=0x2ae054042014) at /work/percona-server/sql/sp_instr.cc:1000
#29 0x000056304dcff611 in sp_lex_instr::reset_lex_and_exec_core (this=0x78b4381ef458, thd=0x78b4380cd3f0, nextp=0x2ae054042014, open_tables=false) at /work/percona-server/sql/sp_instr.cc:443
#30 0x000056304dd0023f in sp_lex_instr::validate_lex_and_execute_core (this=0x78b4381ef458, thd=0x78b4380cd3f0, nextp=0x2ae054042014, open_tables=false) at /work/percona-server/sql/sp_instr.cc:729
#31 0x000056304dd0086b in sp_instr_stmt::execute (this=0x78b4381ef458, thd=0x78b4380cd3f0, nextp=0x2ae054042014) at /work/percona-server/sql/sp_instr.cc:906
#32 0x000056304dcf1f72 in sp_head::execute (this=0x78b4381e1be8, thd=0x78b4380cd3f0, merge_da_on_success=true) at /work/percona-server/sql/sp_head.cc:2223
#33 0x000056304dcf4100 in sp_head::execute_procedure (this=0x78b4381e1be8, thd=0x78b4380cd3f0, args=0x0) at /work/percona-server/sql/sp_head.cc:2875
#34 0x000056304dd79ad1 in Sql_cmd_call::execute_inner (this=0x78b438143b80, thd=0x78b4380cd3f0) at /work/percona-server/sql/sql_call.cc:235
#35 0x000056304df81152 in Sql_cmd_dml::execute (this=0x78b438143b80, thd=0x78b4380cd3f0) at /work/percona-server/sql/sql_select.cc:574
#36 0x000056304def4e76 in mysql_execute_command (thd=0x78b4380cd3f0, first_level=false) at /work/percona-server/sql/sql_parse.cc:4720
#37 0x000056304dd00c59 in sp_instr_stmt::exec_core (this=0x78b438143bf8, thd=0x78b4380cd3f0, nextp=0x2ae054043d84) at /work/percona-server/sql/sp_instr.cc:1000
#38 0x000056304dcff611 in sp_lex_instr::reset_lex_and_exec_core (this=0x78b438143bf8, thd=0x78b4380cd3f0, nextp=0x2ae054043d84, open_tables=false) at /work/percona-server/sql/sp_instr.cc:443
#39 0x000056304dd0023f in sp_lex_instr::validate_lex_and_execute_core (this=0x78b438143bf8, thd=0x78b4380cd3f0, nextp=0x2ae054043d84, open_tables=false) at /work/percona-server/sql/sp_instr.cc:729
#40 0x000056304dd0086b in sp_instr_stmt::execute (this=0x78b438143bf8, thd=0x78b4380cd3f0, nextp=0x2ae054043d84) at /work/percona-server/sql/sp_instr.cc:906
#41 0x000056304dcf1f72 in sp_head::execute (this=0x78b438139cf8, thd=0x78b4380cd3f0, merge_da_on_success=true) at /work/percona-server/sql/sp_head.cc:2223
#42 0x000056304dcf4100 in sp_head::execute_procedure (this=0x78b438139cf8, thd=0x78b4380cd3f0, args=0x78b4380d6960) at /work/percona-server/sql/sp_head.cc:2875
#43 0x000056304dd79ad1 in Sql_cmd_call::execute_inner (this=0x78b4380d6e28, thd=0x78b4380cd3f0) at /work/percona-server/sql/sql_call.cc:235
#44 0x000056304df81152 in Sql_cmd_dml::execute (this=0x78b4380d6e28, thd=0x78b4380cd3f0) at /work/percona-server/sql/sql_select.cc:574
#45 0x000056304def4e76 in mysql_execute_command (thd=0x78b4380cd3f0, first_level=true) at /work/percona-server/sql/sql_parse.cc:4720
#46 0x000056304def7014 in dispatch_sql_command (thd=0x78b4380cd3f0, parser_state=0x2ae054045aa0, update_userstat=false) at /work/percona-server/sql/sql_parse.cc:5325
#47 0x000056304deec4f6 in dispatch_command (thd=0x78b4380cd3f0, com_data=0x2ae054046b70, command=COM_QUERY) at /work/percona-server/sql/sql_parse.cc:1956
#48 0x000056304deea868 in do_command (thd=0x78b4380cd3f0) at /work/percona-server/sql/sql_parse.cc:1404
#49 0x000056304e1145d6 in handle_connection (arg=0x23fb4af16320) at /work/percona-server/sql/conn_handler/connection_handler_per_thread.cc:307
#50 0x000056304fba1862 in pfs_spawn_thread (arg=0x23fb4aea0760) at /work/percona-server/storage/perfschema/pfs.cc:2899
#51 0x00003c74345ce609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#52 0x0000038f69fe7293 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```
</details>

<details><summary>`*mdl_request` value at `MDL_context::acquire_lock()` (frame 7)</summary>

```
(gdb) p *mdl_request
$4 = {type = MDL_SHARED_READ, duration = MDL_TRANSACTION, next_in_list = 0x0, prev_in_list = 0x0, ticket = 0x0, key = {m_length = 15, m_db_name_length = 5, m_object_name_length = 7, m_ptr = "\004mysql\000columns", '\000' <repeats 372 times>, static m_namespace_to_wait_state_name = {{m_key = 102, m_name = 0x56304b210646 "Waiting for global read lock", m_flags = 0, m_documentation = 0x56304b20ff20 ""}, {m_key = 103, m_name = 0x56304b210663 "Waiting for backup lock", m_flags = 0, m_documentation = 0x56304b20ff20 ""}, {m_key = 104, m_name = 0x56304b210680 "Waiting for tablespace metadata lock", m_flags = 0, m_documentation = 0x56304b20ff20 ""}, {m_key = 105, m_name = 0x56304b2106a8 "Waiting for schema metadata lock", m_flags = 0, m_documentation = 0x56304b20ff20 ""}, {m_key = 106, m_name = 0x56304b2106d0 "Waiting for table metadata lock", m_flags = 0, m_documentation = 0x56304b20ff20 ""}, {m_key = 107, m_name = 0x56304b2106f0 "Waiting for stored function metadata lock", m_flags = 0, m_documentation = 0x56304b20ff20 ""}, {m_key = 108, m_name = 0x56304b210720 "Waiting for stored procedure metadata lock", m_flags = 0, m_documentation = 0x56304b20ff20 ""}, {m_key = 109, m_name = 0x56304b210750 "Waiting for trigger metadata lock", m_flags = 0, m_documentation = 0x56304b20ff20 ""}, {m_key = 110, m_name = 0x56304b210778 "Waiting for event metadata lock", m_flags = 0, m_documentation = 0x56304b20ff20 ""}, {m_key = 111, m_name = 0x56304b210798 "Waiting for commit lock", m_flags = 0, m_documentation = 0x56304b20ff20 ""}, {m_key = 112, m_name = 0x56304b2107b0 "User lock", m_flags = 0, m_documentation = 0x56304b20ff20 ""}, {m_key = 113, m_name = 0x56304b2107c0 "Waiting for locking service lock", m_flags = 0, m_documentation = 0x56304b20ff20 ""}, {m_key = 114, m_name = 0x56304b2107e8 "Waiting for spatial reference system lock", m_flags = 0, m_documentation = 0x56304b20ff20 ""}, {m_key = 115, m_name = 0x56304b210812 "Waiting for acl cache lock", m_flags = 0, m_documentation = 0x56304b20ff20 ""}, {m_key = 116, m_name = 0x56304b210830 "Waiting for column statistics lock", m_flags = 0, m_documentation = 0x56304b20ff20 ""}, {m_key = 117, m_name = 0x56304b210858 "Waiting for resource groups metadata lock", m_flags = 0, m_documentation = 0x56304b20ff20 ""}, {m_key = 118, m_name = 0x56304b210888 "Waiting for foreign key metadata lock", m_flags = 0, m_documentation = 0x56304b20ff20 ""}, {m_key = 119, m_name = 0x56304b2108b0 "Waiting for check constraint metadata lock", m_flags = 0, m_documentation = 0x56304b20ff20 ""}, {m_key = 120, m_name = 0x56304b2108db "Waiting for table backup lock", m_flags = 0, m_documentation = 0x56304b20ff20 ""}}}, m_src_file = 0x56304b190c20 "/work/percona-server/sql/table.h", m_src_line = 2831}
(gdb) p mdl_request.key.m_ptr
$5 = "\004mysql\000columns", '\000' <repeats 372 times>
```
</details>

<details><summary>Stacktrace at the point where MDL_EXCLUSIVE, MDL_TRANSACTION lock was previously asked for `mysql.columns`:</summary>

```
#0  MDL_context::acquire_lock (this=0x235454000c58, mdl_request=0x74707a286290, lock_wait_timeout=31536000) at /work/percona-server/sql/mdl.cc:3375
#1  0x000056304db9879d in MDL_context::upgrade_shared_lock (this=0x235454000c58, mdl_ticket=0x23545401e480, new_type=MDL_EXCLUSIVE, lock_wait_timeout=31536000) at /work/percona-server/sql/mdl.cc:3772
#2  0x000056304e017228 in upgrade_lock_for_tables_in_tablespace (thd=0x235454000be0, table_mdl_reqs=0x74707a286570) at /work/percona-server/sql/sql_tablespace.cc:904
#3  0x000056304e017995 in Sql_cmd_alter_tablespace::execute (this=0x23545400be58, thd=0x235454000be0) at /work/percona-server/sql/sql_tablespace.cc:1026
#4  0x000056304def4e76 in mysql_execute_command (thd=0x235454000be0, first_level=false) at /work/percona-server/sql/sql_parse.cc:4720
#5  0x000056304df43658 in (anonymous namespace)::Execute_sql_statement::execute_server_code (this=0x74707a287f10, thd=0x235454000be0) at /work/percona-server/sql/sql_prepare.cc:2284
#6  0x000056304df45c98 in Prepared_statement::execute_server_runnable (this=0x74707a287bd0, server_runnable=0x74707a287f10) at /work/percona-server/sql/sql_prepare.cc:3125
#7  0x000056304df4735d in Ed_connection::execute_direct (this=0x74707a287f90, server_runnable=0x74707a287f10) at /work/percona-server/sql/sql_prepare.cc:3669
#8  0x000056304df4720d in Ed_connection::execute_direct (this=0x74707a287f90, sql_text=...) at /work/percona-server/sql/sql_prepare.cc:3643
#9  0x000056304f07c210 in dd::execute_query (thd=0x235454000be0, q_buf="ALTER TABLESPACE mysql ENCRYPTION = 'Y'") at /work/percona-server/sql/dd/impl/utils.cc:47
#10 0x000056304f064146 in dd::alter_tablespace_encryption (thd=0x235454000be0, tablespace_name=0x7ffcca756ad0 "mysql", encryption=true) at /work/percona-server/sql/dd/impl/dictionary_impl.cc:779
#11 0x000056304f60d8c8 in resume_alter_encrypt_tablespace (thd=0x235454000be0) at /work/percona-server/storage/innobase/fsp/fsp0fsp.cc:4779
#12 0x000056304f60ddda in fsp_init_resume_alter_encrypt_tablespace () at /work/percona-server/storage/innobase/fsp/fsp0fsp.cc:4831
#13 0x000056304f7276ac in std::__invoke_impl<void, void (*&)()> (__f=@0x74707a288bc0: 0x56304f60dd9f <fsp_init_resume_alter_encrypt_tablespace()>) at /usr/include/c++/9/bits/invoke.h:60
#14 0x000056304f727662 in std::__invoke<void (*&)()> (__fn=@0x74707a288bc0: 0x56304f60dd9f <fsp_init_resume_alter_encrypt_tablespace()>) at /usr/include/c++/9/bits/invoke.h:95
#15 0x000056304f72762c in std::_Bind<void (*())()>::__call<void>(std::tuple<>&&, std::_Index_tuple<>) (this=0x74707a288bc0, __args=...) at /usr/include/c++/9/functional:400
#16 0x000056304f7275b5 in std::_Bind<void (*())()>::operator()<, void>() (this=0x74707a288bc0) at /usr/include/c++/9/functional:484
#17 0x000056304f72750d in Runnable::operator()<void (*)()> (this=0x563052df9aa0, f=@0x563052df9a98: 0x56304f60dd9f <fsp_init_resume_alter_encrypt_tablespace()>) at /work/percona-server/storage/innobase/include/os0thread-create.h:101
#18 0x000056304f727402 in std::__invoke_impl<void, Runnable, void (*)()> (__f=...) at /usr/include/c++/9/bits/invoke.h:60
#19 0x000056304f72735d in std::__invoke<Runnable, void (*)()> (__fn=...) at /usr/include/c++/9/bits/invoke.h:95
#20 0x000056304f7272ad in std::thread::_Invoker<std::tuple<Runnable, void (*)()> >::_M_invoke<0ul, 1ul> (this=0x563052df9a98) at /usr/include/c++/9/thread:244
#21 0x000056304f72724f in std::thread::_Invoker<std::tuple<Runnable, void (*)()> >::operator() (this=0x563052df9a98) at /usr/include/c++/9/thread:251
#22 0x000056304f727220 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<Runnable, void (*)()> > >::_M_run (this=0x563052df9a90) at /usr/include/c++/9/thread:195
#23 0x000062ca243b6de4 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#24 0x00003c74345ce609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#25 0x0000038f69fe7293 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```
</details>

Reply:
> fix mysql tablesapce encryption-decryption before we accept connections

Do you mean to change the code that resumes encryption at the server startup? I believe it won't fix the issue. While the supplied test restarts the server, restarting is not a requirement.

It's is possible to run two clients one of which encrypts-decrypts "mysql" tablespace in a cycle, while the other one runs DDL statements in a loop. Here is testcase if you want to try it locally:

<details><summary>testcase.test</summary>

```
--connect (con1,localhost,root,,)
--connect (con2,localhost,root,,)

--connection default

DELIMITER //;

CREATE PROCEDURE my_infinite_alter_ts()
BEGIN
  WHILE TRUE DO
    ALTER TABLESPACE mysql ENCRYPTION='Y';
    ALTER TABLESPACE mysql ENCRYPTION='N';
  END WHILE;
END//

CREATE PROCEDURE infinite_check_warnings()
BEGIN
  WHILE TRUE DO
    CREATE TABLE IF NOT EXISTS mtr.error_log (
      id INT AUTO_INCREMENT PRIMARY KEY,
      suspicious INT DEFAULT 1,
      line VARCHAR(10000) CHARSET latin1 DEFAULT NULL
    );
    CALL mtr.check_warnings(@result);
  END WHILE;
END//

DELIMITER ;//

--connection con1
--send CALL my_infinite_alter_ts()

--connection con2

CALL infinite_check_warnings();

--connection default

--send_quit con1
--send_quit con2

sleep 3;
DROP PROCEDURE my_infinite_alter_ts;
DROP PROCEDURE infinite_check_warnings;

--disconnect con1
--disconnect con2
```
</details>

Requires similar `.opt` file as in this pull request.
It's probably enough to create and destroy tables in a loop, but I did not try exactly that. Instead, the test calls `mtr.check_warnings()` procedure in a loop. That definitely triggers the assertion.

You may want to increase test timeouts, as this test may take a long time.

Reply:
> Can we fix this issue as separate PR?

If it's possible, I'd like to address all comments while having everything in this PR first, to reduce confusion. And after that I'll just separate the part into another PR. Otherwise, I have no specific preferences.

Reply:
I looked into the testcase provided by you and I am able to reproduce the issue. So it looks like a deadlock of opening DD tables.
A: ALTER TABLESPCE mysql is a bit special. It is trying to open all tables inside it. This is for adding encryption='y' attribute to table object for example opens tables in this order: a->b->c
B: Other DD operations that open DD tables. using Attachable operation (for example: opens c->b-a)

This is the reason for the deadlock. Initially, I thought it had to do with tablespace lock on MySQL tablespace. Not the case.

Please fix the lock ordering to avoid deadlocks and remove the connection blocking logic. Also, connection blocking solution wouldn't work for existing connections, that would still access DD and cause deadlocks.

Please create separate PR for this issue

Reply:
> separate PR for this issue

Fixing this issue caused a data race in InnoDB's code to surface, so there are two pull requests:
https://github.com/percona/percona-server/pull/4707 (this one is about InnoDB's `fil_io_set_encryption()`)
https://github.com/percona/percona-server/pull/4708 (this one is about MDL ordering for ALTER TABLESPACE)